R/FilterStateChoices.R

```diff
@@ -448,63 +448,92 @@ ChoicesFilterState <- R6::R6Class( # nolint
+                )
                 teal.widgets::updateOptionalSelectInput(
-                  session, "selection",
+                  session = session,
+                  inputId = "selection",
+                  choices = stats::setNames(private$get_choices(), labels),
                   selected = private$get_selected()
                 )
-                return(NULL)
-              } else {
-                input$selection
               }
+              NULL
+            })
+          })
+
+          if (private$is_checkboxgroup()) {
+            private$observers$selection <- observeEvent(
+              ignoreNULL = FALSE,
+              ignoreInit = TRUE, # ignoreInit: should not matter because we set the UI with the desired initial state
+              eventExpr = input$selection,
+              handlerExpr = {
+                logger::log_trace("ChoicesFilterState$server_inputs@2 changed selection, id: { private$get_id() }")
+
+                selection <- if (is.null(input$selection) && private$is_multiple()) {
+                  character(0)
+                } else {
+                  input$selection
+                }
+
+                private$set_selected(selection)
+              }
+            )
+          } else {
+            private$observers$selection <- observeEvent(
+              ignoreNULL = FALSE,
+              ignoreInit = TRUE, # ignoreInit: should not matter because we set the UI with the desired initial state
+              eventExpr = input$selection_open, # observe click on a dropdown
+              handlerExpr = {
+                if (!isTRUE(input$selection_open)) { # only when the dropdown got closed
+                  logger::log_trace("ChoicesFilterState$server_inputs@2 changed selection, id: { private$get_id() }")
+
+                  selection <- if (is.null(input$selection) && private$is_multiple()) {
+                    character(0)
+                  } else if (isTRUE(length(input$selection) != 1) && !private$is_multiple()) {
+                    # In optionalSelectInput user is able to select mutliple options. But if FilterState is not multiple
+                    # we should prevent this selection to be processed further.
+                    # This is why notification is thrown and dropdown is changed back to latest selected.
+                    showNotification(paste(
+                      "This filter exclusively supports single selection.",
+                      "Any additional choices made will be disregarded."
+                    ))
+                    teal.widgets::updateOptionalSelectInput(
+                      session, "selection",
+                      selected = private$get_selected()
+                    )
+                    return(NULL)
+                  } else {
+                    input$selection
+                  }
+                  private$set_selected(selection)
+                }
+              }
+            )
+          }
```

Comment:
@chlebowa @m7pr I needed to split one `observeEvent(input$selection`, on two:
```
if (private$is_checkbox()) {
  observeEvent(input$selection, ...)
} else {
  observeEvent(input$selection_open, ...)
}
```

We have to block optionalSelectInput (picker) to be triggered on every selection change - see SO issue https://stackoverflow.com/questions/68845540/prevent-pickerinput-from-updating-every-time-something-is-selected-r-shiny

So now, pickerInput selection is triggering futher processes only when user closes dropdown.